### DISCUSSION TITLE
Compare series of images

### DISCUSSION THREAD

**Discussion by bkenobi** [ImageMagick/ImageMagick#3940] [Help] [answered]
I have a series of images that is from the output of an analysis with a fixed background (flow through an unstable nozzle).  There is a portion of the picture that is moving due to the physics of the model.  I was hoping to use the compare tool to see the area where things change over the series of images (iow, where does the flow spread as a maximum).  I have 1000 images per animation and several animations so it's not exactly simple to manually find the pictures with the maximum spread.  I found the guide for doing this for two images but I'm not sure how to scale it to 3 let alone 1000.

I am using ImageMagick 6.8.8-1 Q16 x86_64 2021-02-23 on Linux

https://imagemagick.org/script/compare.php

![file1](https://user-images.githubusercontent.com/6903751/125975983-d797510d-4c7a-42dd-bbd7-1ceb3e1e2133.jpg)
![file2](https://user-images.githubusercontent.com/6903751/125975987-7898a364-95d3-40a4-ab9c-ad24eb89e76c.jpg)
![file3](https://user-images.githubusercontent.com/6903751/125975990-467d73cd-6d8a-419d-aa89-138a7cd07e7b.jpg)

Just continuing to brainstorm, the other option would be to delete/make transparent the blue background color and add the images together. I've only used ImageMagick for basic things (animation creation, viewing images/animations) and one time creating a composite animation. I know how powerful it is but get started is intimidating!

**Comment by fmw42** [comment]
compare can only do two images at a time.

**Reply by fmw42** [reply]
I suggest you post two or 3 images so we can see what you have.

**Reply by bkenobi** [reply]
I assumed that would be the case, but wasn't sure if there was a cookbook example of how to expand via a script etc to expand that.  I can write a script given a basic set of commands, I'm just not sure of an approach that I could use.  IOW, do I:

1. compare the image1 to image2 and save diff12
2. compare image1 to image3 and save diff13
3. compare diff12 to diff13
4. etc?

I'm just not sure how to approach the process.

**Comment by snibgo** [comment]
What is your platform? I suspect you will need a shell script for this.

> (iow, where does the flow spread as a maximum)

Without seeing samples, I can't advise. What does a maximum spread look like?

**Reply by bkenobi** [reply]
I'm running the command on Linux.  I'll post a few images.  I need to crop a few pixel lines off the bottom and am rusty on the -crop command.  Working on it.

**Comment by fmw42** [comment, marked as answer]
As compare is very slow right now (stay tuned), I would suggest using -compose difference -composite to get the abs difference between frames. It would be much faster.

**Reply by bkenobi** [reply]
This is good to know considering how many images I need to process.

**Reply by bkenobi** [reply]
I wrote a script that changes the blue background to transparent and then composites all images.  It's close enough for what I need.

**Comment by snibgo** [comment]
Taking your three posted images:
```
f:\web\im>%IMG7%magick compare -metric RMSE nozzle1.jpg nozzle2.jpg NULL:
1137.6 (0.0173586)

f:\web\im>%IMG7%magick compare -metric RMSE nozzle1.jpg nozzle3.jpg NULL:
6390.46 (0.0975122)
```
The numbers in parentheses are on a scale of 0.0 to 1.0. Larger means more change.

A simple comparisons shows that 1 and 3 differ by far more than the difference between 1 and 2. There is a clear factor of six.

I don't know what you actually want to measure. Perhaps the change between successive frames.

**Reply by bkenobi** [reply]
I used composite to make a comparison through a script.  Clues provided here let me to the right commands to get what I need.
